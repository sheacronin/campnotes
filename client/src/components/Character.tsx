import React, { useState } from 'react';
import { ICharacter } from '../types';
import EditCharacter from './EditCharacter';

interface CharacterProps {
  character: ICharacter;
  setCharacters: Function;
}

function Character({ character, setCharacters }: CharacterProps) {
  const [isEditing, setIsEditing] = useState(false);

  async function deleteCharacter(id: number) {
    try {
      const deletedCharacter = await fetch(
        `http://localhost:5000/characters/${id}`,
        { method: 'DELETE' }
      );
      console.log(deletedCharacter);
      setCharacters((prevState: ICharacter[]) =>
        prevState.filter((character) => character.id !== id)
      );
    } catch (error) {
      console.error(error);
    }
  }

  return isEditing ? (
    <EditCharacter
      character={character}
      setIsEditing={setIsEditing}
      setCharacters={setCharacters}
    />
  ) : (
    <article className="character">
      <h3>{character.name}</h3>
      <div>
        {character.alignment} | {character.race}
      </div>
      <div>{character.description}</div>
      <button onClick={() => deleteCharacter(character.id)}>Delete</button>
      <button onClick={() => setIsEditing(true)}>Edit</button>
    </article>
  );
}

export default Character;
