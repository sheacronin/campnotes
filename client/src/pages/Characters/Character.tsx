import React from 'react';
import { ICharacter } from './useCharacters';
import EditCharacter from './EditCharacter';
import useEditCharacter from './useEditCharacter';

interface CharacterProps {
  character: ICharacter;
  deleteCharacter: Function;
  addCharacter: Function;
  editCharacter: Function;
}

function Character({
  character,
  deleteCharacter,
  addCharacter,
  editCharacter,
}: CharacterProps) {
  const { isEditing, onSubmitForm, openEditForm, closeEditForm } =
    useEditCharacter(addCharacter, editCharacter);

  return isEditing ? (
    <EditCharacter
      character={character}
      onSubmitForm={onSubmitForm}
      closeEditForm={closeEditForm}
    />
  ) : (
    <article className="character">
      <h3>{character.name}</h3>
      <div>
        {character.alignment} | {character.race}
      </div>
      <div>{character.description}</div>
      <button onClick={() => deleteCharacter(character.id)}>Delete</button>
      <button onClick={() => openEditForm()}>Edit</button>
    </article>
  );
}

export default Character;
