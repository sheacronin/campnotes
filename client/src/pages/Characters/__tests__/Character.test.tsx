import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Character from '../Character';

describe('Character component', () => {
  const CHARACTER = {
    id: 1,
    name: 'Name',
    description: 'Desc',
    race: 'Race',
    alignment: 'NG',
  };
  const deleteCharacter = jest.fn();
  const addCharacter = jest.fn();
  const editCharacter = jest.fn();

  it('renders correct heading', () => {
    render(
      <Character
        character={CHARACTER}
        deleteCharacter={deleteCharacter}
        addCharacter={addCharacter}
        editCharacter={editCharacter}
      />
    );
    expect(screen.getByRole('heading').textContent).toMatch(/Name/);
  });

  it('matches snapshot', () => {
    const { container } = render(
      <Character
        character={CHARACTER}
        deleteCharacter={deleteCharacter}
        addCharacter={addCharacter}
        editCharacter={editCharacter}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('calls deleteCharacter after delete button click', () => {
    render(
      <Character
        character={CHARACTER}
        deleteCharacter={deleteCharacter}
        addCharacter={addCharacter}
        editCharacter={editCharacter}
      />
    );
    const deleteButton = screen.getByRole('button', { name: 'Delete' });
    userEvent.click(deleteButton);
    expect(deleteCharacter).toHaveBeenCalled();
  });
});
