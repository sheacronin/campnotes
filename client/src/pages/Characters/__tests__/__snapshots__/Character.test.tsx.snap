// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character component matches snapshot 1`] = `
<div>
  <article
    class="character"
  >
    <h3>
      Name
    </h3>
    <div>
      NG
       | 
      Race
    </div>
    <div>
      Desc
    </div>
    <button>
      Delete
    </button>
    <button>
      Edit
    </button>
  </article>
</div>
`;
